openmc/statepoint.py

```diff
@@ -94,475 +91,385 @@ def __init__(self, filename):
-            for tally_key in self._tally_keys:
+            # Iterate over all Tallies
+            for tally_key in tally_keys:
 
-                # Get this Tally
-                tally = self._tallies[tally_key]
+                # Read the Tally size specifications
+                n_realizations = self._f['{0}{1}/n_realizations'.format(base, tally_key)].value
 
-                # Compute the total number of bins for this Tally
-                num_tot_bins = tally.num_bins
+                # Create Tally object and assign basic properties
+                tally = openmc.Tally(tally_key)
+                tally._statepoint = self
+                tally.estimator = self._f['{0}{1}/estimator'.format(
+                    base, tally_key)].value.decode()
+                tally.num_realizations = n_realizations
 
-                # Extract Tally data from the file
-                data = self._f['{0}{1}/results'.format(base, tally_key)].value
-                sum = data['sum']
-                sum_sq = data['sum_sq']
+                # Read the number of Filters
+                n_filters = self._f['{0}{1}/n_filters'.format(base, tally_key)].value
 
-                # Define a routine to convert 0 to 1
-                def nonzero(val):
-                    return 1 if not val else val
+                subbase = '{0}{1}/filter '.format(base, tally_key)
 
-                # Reshape the results arrays
-                new_shape = (nonzero(tally.num_filter_bins),
-                             nonzero(tally.num_nuclides),
-                             nonzero(tally.num_score_bins))
+                # Initialize all Filters
+                for j in range(1, n_filters+1):
 
-                sum = np.reshape(sum, new_shape)
-                sum_sq = np.reshape(sum_sq, new_shape)
+                    # Read the Filter type
+                    filter_type = self._f['{0}{1}/type'.format(subbase, j)].value.decode()
 
-                # Set the data for this Tally
-                tally.sum = sum
-                tally.sum_sq = sum_sq
+                    # Read the Filter offset
+                    offset = self._f['{0}{1}/offset'.format(subbase, j)].value
 
-        # Indicate that Tally results have been read
-        self._results = True
+                    n_bins = self._f['{0}{1}/n_bins'.format(subbase, j)].value
 
-    def read_source(self):
-        """Read and store source sites from the statepoint file. By default, source
-        sites are not loaded upon initialization.
+                    if n_bins <= 0:
+                        msg = 'Unable to create Filter "{0}" for Tally ID="{1}" ' \
+                              'since no bins were specified'.format(j, tally_key)
+                        raise ValueError(msg)
 
-        """
+                    # Read the bin values
+                    if filter_type in ['energy', 'energyout']:
+                        bins = self._f['{0}{1}/bins'.format(subbase, j)].value
```

Comment:
It looks like the expressions in each of the ``if-elif-else`` code blocks are the same and can be simplified to a single line of code without need for the conditional.